cvat/apps/engine/ddln/tasks/vls_lines/persistence/cvat.py

```diff
@@ -0,0 +1,209 @@
+        "group": 0,
+        "z_order": 0,
+        "occluded": False,
+        "type": "rays",
+        "label": "Horizontal line",
+    }
+    writer._annotations.add_shape(writer._annotations.LabeledShape(**lon_shape))
+    writer._annotations.add_shape(writer._annotations.LabeledShape(**lat_shape))
+
+
+def _parse_rays(lon, lat, reporter):
+    lon_shape, lon_attrs = lon
+    runway_id = lon_attrs['Runway_ID']
+    left_visible = bool(int(lon_attrs['First_line(1)']))
+    right_visible = bool(int(lon_attrs['Second_line(2)']))
+    center_visible = bool(int(lon_attrs['Third_line(3)']))
+    left, right, center = _parse_lines(lon_shape)
+    if lat:
+        lat_shape, lat_attrs = lat
+        start_visible = bool(int(lat_attrs['First_line(1)']))
+        designator_visible = bool(int(lat_attrs['Second_line(2)']))
+        end_visible = bool(int(lat_attrs['Third_line(3)']))
+        start, designator, end = _parse_lines(lat_shape)
+    else:
+        start_visible = end_visible = designator_visible = False
+        start = designator = end = None
+
+    runway = Runway(runway_id, left, right, center, start, end, designator)
+    runway.calculate_vanishing_points(reporter)
+    runway.fix_order(reporter)
+    runway.apply_visibility(left_visible, right_visible, center_visible, start_visible, end_visible, designator_visible)
+    return runway
+
+
+def _parse_lines(shape):
+    points = (Point(x, y) for x, y in grouper(shape.points, 2))
+    lines = [Line.by_two_points(a, b) for a, b in grouper(points, 2)]
+    if len(lines) < 3:
+        raise RunwayParseError("Not enough lines (should be at least 3)")
+    first, second, third, *rest = lines
+    return first, second, third
+
+
+def _get_points_list(runway):
+    # Coordinates for non-visible lines are not stored in csv,
+    # so have to guess them
+    lon_lines = _guess_lines(runway.lon_vanishing_point, runway.left_line, runway.center_line, runway.right_line)
+    lat_lines = _guess_lines(runway.lat_vanishing_point, runway.start_line, runway.designator_line, runway.end_line)
+    intersections = []
+    for a in lon_lines:
+        row = []
+        for b in lat_lines:
+            row.append(a.intersect(b))
+        intersections.append(row)
+
+    lon_points = []
+    for lon_index in range(3):
+        lon_points.append(intersections[lon_index][0])
+        lon_points.append(intersections[lon_index][2])
+    lat_points = []
+    for lat_index in range(3):
+        lat_points.append(intersections[0][lat_index])
+        lat_points.append(intersections[2][lat_index])
+
+    lon_result = []
+    for point in lon_points:
+        lon_result.append(point.x)
+        lon_result.append(point.y)
+    lat_result = []
+    for point in lat_points:
+        lat_result.append(point.x)
+        lat_result.append(point.y)
+    return lon_result, lat_result
+
+
+def _guess_lines(vanishing_point, first, second, third):
+    missing_lines_amount = sum(line is None for line in [first, second, third])
```

Comment:
can you explain the purpose of this and above functions?

Reply:
When I implement dump feature, I also implement upload feature in order to be able test dump feature and in order to be able handle errors in dump if they happen. But dump into required output format is lossy, the information about line segments edges coordinates are lost, as well as values of invisible lines. So I have to guess them in upload.

Upload is not used in annotators workflow and mainly used by me.